### PR TITLE
Support pom-only dependencies

### PR DESCRIPTION
**Pull request by jin**
Fixes https://github.com/bazelbuild/rules_jvm_external/issues/98

Add a whitelist of artifacts that are "aggregators", that is, they only have a POM file and no jar. 

If Coursier tells us that a particular artifact doesn't have a file, we generate a `java_library` instead, and re-export its deps.

### CODE REVIEW

File: coursier.bzl [hunk 1]
```diff
@@ -101,9 +108,6 @@ def generate_imports(repository_ctx, dep_tree, srcs_dep_tree = None):
     # The list of java_import/aar_import declaration strings to be joined at the end
     all_imports = []
 
-    # A mapping of FQN to the artifact's sha256 checksum
```
Comment (jin) [inline coursier.bzl]: drive-by removal of the checksumming code

File: coursier.bzl [hunk 2]
```diff
@@ -125,29 +129,17 @@ def generate_imports(repository_ctx, dep_tree, srcs_dep_tree = None):
...
                     # If not, it's a relative path to the one in output_base/external/$maven/v1/...
                     artifact_relative_path = _normalize_to_unix_path(artifact_path)
-                target_label = _escape(_strip_packaging_and_classifier(artifact["coord"]))
+                target_label = _escape(_strip_packaging_classifier(artifact["coord"]))
                 srcjar_paths[target_label] = artifact_relative_path
 
     # Iterate through the list of artifacts, and generate the target declaration strings.
     for artifact in dep_tree["dependencies"]:
         artifact_path = artifact["file"]
+        target_label = _escape(_strip_packaging_classifier(artifact["coord"]))
```
Comment (jin) [inline coursier.bzl]: instead of using the artifact path as the key to dedupe target generation, we use the target label as the key, since it's canonical and unique anyway

File: coursier.bzl [hunk 3]
```diff
@@ -31,12 +34,16 @@ load("@{repository_name}//:jvm_import.bzl", "jvm_import")
...
+        coord = coord.replace(":%s:" % packaging_type, ":")
+    for classifier_type in ["sources", "natives"]:
+        coord = coord.replace(":%s:" % classifier_type, ":")
+
+    return coord
 
-def _strip_packaging_and_classifier_and_version(coord):
-    return ":".join(_strip_packaging_and_classifier(coord).split(":")[:-1])
+def _strip_packaging_classifier_version(coord):
+    return ":".join(_strip_packaging_classifier(coord).split(":")[:-1])
```
Comment (dkelmer) [inline coursier.bzl]: Why the change to the name? _strip_packaging_and_classifier is clearer (without the and it sounds like you're stripping a single thing called "packaging classifier"). Also with the version method

File: coursier.bzl [hunk 4]
```diff
@@ -31,12 +34,16 @@ load("@{repository_name}//:jvm_import.bzl", "jvm_import")
 # download them if it's not asked to to resolve "eclipse-plugin".
 _COURSIER_PACKAGING_TYPES = ["jar", "aar", "bundle", "eclipse-plugin"]
 
-# Super hacky :(
-def _strip_packaging_and_classifier(coord):
-    return coord.replace(":jar:", ":").replace(":aar:", ":").replace(":sources:", ":")
+def _strip_packaging_classifier(coord):
+    for packaging_type in _COURSIER_PACKAGING_TYPES + ["pom"]:
```
Comment (dkelmer) [inline coursier.bzl]: Add "pom" to _COURSIER_PACKAGING_TYPES instead of adding it here

File: private/special_artifacts.bzl [hunk 5]
```diff
@@ -0,0 +1,3 @@
+POM_ONLY_ARTIFACTS = {
+    "com.github.fommil.netlib:all": True,
```
Comment (dkelmer) [inline private/special_artifacts.bzl]: Does this mean that we need to hardcode all pom only artifacts?

File: coursier.bzl [hunk 6]
```diff
@@ -263,7 +278,7 @@ def generate_imports(repository_ctx, dep_tree, srcs_dep_tree = None):
             reverse_dep_coords = [reverse_dep["coord"] for reverse_dep in reverse_deps]
             reverse_dep_pom_paths = [
                 repository_ctx.path(reverse_dep["file"].replace(".jar", ".pom").replace(".aar", ".pom"))
-                for reverse_dep in reverse_deps
+                for reverse_dep in reverse_deps if reverse_dep["file"]
```
Comment (dkelmer) [inline coursier.bzl]: I think the "if reverse_dep["file"] part isn't necessary because the previous elif clause will catch the pom only dependencies which are the ones that don't have a file

File: private/special_artifacts.bzl (hunk 5 above)
Comment (jin) [reply private/special_artifacts.bzl]: Yes. My estimate is that this is a rare occurrence, so we probably shouldn't bake huge custom logic into the rule.

File: coursier.bzl (hunk 3 above)
Comment (jin) [reply coursier.bzl]: Reverted.

File: coursier.bzl (hunk 4 above)
Comment (jin) [reply coursier.bzl]: _COURSIER_PACKAGING_TYPES is fed into the coursier CLI, which doesn't recognize `pom` as a packaging type. I'll add a comment about it.

File: coursier.bzl (hunk 6 above)
Comment (jin) [reply coursier.bzl]: Great catch. Done.
